### ISSUE TITLE
About contrastive loss in resnet.py

### ISSUE THREAD

**Issue by BaophanN** [tobysunsun/deeperforward#1]
Thanks author for this novel work, but I have 1 concern. In the update function of resnet.py. 
```
    def update(self, dataloader):
        devices = [next(layer.parameters()).device for layer in self.layers]

        shortcut = torch.zeros(1)
        criterion = nn.CrossEntropyLoss()
        # add InfoNCELoss for representation learning
        contrastive_loss = InfoNCELoss()
        self.train()
        for x, labels in dataloader:
            x, labels = x.to(devices[0]), labels.to(devices[0])
            x = F.layer_norm(x, x.shape[1:])
            for i, layers in enumerate(self.layers):
                x, labels = x.to(devices[i]), labels.to(devices[i])

                self.optimizers[i].zero_grad()
                x, g = layers(x)
                # y, g = layers(x)  # memory saving
                loss = criterion(g, labels)
                loss.backward()
                # alloc_memory = torch.cuda.memory_allocated(device=devices[0])
                self.optimizers[i].step()

                # before_memory = torch.cuda.memory_allocated(device=devices[0])
                # with torch.no_grad(): # memory-saving
                #     del x, g, loss
                #     torch.cuda.empty_cache()
                # after_memory = torch.cuda.memory_allocated(device=devices[0])
                # print(f"Memory: {before_memory / 1024**2:.2f} -> {after_memory/ 1024**2:.2f}")
                # x = y

                # reversed_memory = torch.cuda.memory_reserved(device=devices[0])
                # print(f'Allocated Memory: {alloc_memory / 1024 ** 2:.2f} MB', f' Reserved Memory: {reversed_memory / 1024 ** 2:.2f} MB')

                if self.shortcut_flag[i] == ADD_SHORTCUT:
                    x += shortcut.to(x.device)
                elif self.shortcut_flag[i] == CONCAT_SHORTCUT:
                    x = torch.cat([x, shortcut.to(x.device)], dim=1)

                if self.input_shortcut_flag[i]:
                    shortcut = F.avg_pool2d(x, 2, stride=2).detach() if self.downsample_flag[i] else x.detach()

        for scheduler in self.schedulers:
            scheduler.step()
```
You create a variable contrastive_loss, but I see that this is never used in the training. The training actually uses the cross entropy loss. Is this a typo in the function or what is your intention of doing this. Thank you.

**Comment by BaophanN** [comment]
@tobysunsun Hi author, can you please answer my question. Thank you so much.

**Comment by tobysunsun** [comment]
Thanks for your question. The InfoNCE loss function is for our further research. The current version does not use InfoNCE, but only uses the cross entropy loss, as in the paper. Sorry for confusing you.

**Comment by BaophanN** [comment]
Thanks for your response, so the InfoNCE loss function comes from an unpublished work or your current research, doesn't it?

**Comment by tobysunsun** [comment]
Yes, combining with InfoNCE loss or other representation learning methods is still under research.
